File: code/worker_thread/blog/python/prodcons.py
#!/usr/bin/env python

from random import randint
from time import sleep
from functools import partial
from threading import Thread
from worker_thread import WorkerThread

def mul_two(i):
    print('mul_two({}) is {}'.format(i, i*2))

def consume(i):
    print('consume with data: {}'.format(i))
    mul_two(i)
    sleep(randint(2, 5))

def producer(worker_thread, loops):
    for i in range(loops):
        print('pushTask with data: {}'.format(i))
        worker_thread.putTask(partial(consume, i))
        sleep(randint(1, 3))

def main():
    nloops = randint(2, 5)

    myworker = WorkerThread("worker")
    myworker.start()

    myproducer = Thread(target=producer, args=(myworker, nloops))
    myproducer.start()
    myproducer.join()

    myworker.stop()
    myworker.join()

    print('all DONE')

if __name__ == '__main__':
    main()

File: code/worker_thread/blog/cxx/prodcons.cpp
#include <iostream>
#include <thread>
#include <chrono>

#include "random.hpp"
#include "worker_thread.hpp"

void mul_two(int i) {
    std::cout << "mul_two(" << i << ") is " << i*2 << std::endl;
}

void consume(int i) {
    mul_two(i);
    std::this_thread::sleep_for(std::chrono::seconds(randint(2, 5)));
}

void producer(std::shared_ptr<TaskQueue> queue, int loops) {
    for (int i = 0; i < loops; i++) {
        queue->PushTask(&consume, i);
        std::this_thread::sleep_for(std::chrono::seconds(randint(1, 3)));
    }
}

int main() {
    int nloops = randint(2, 5);

    WorkerThread myworker("worker");
    myworker.Start();

    std::thread myproducer(&producer, myworker.GetTaskQueue(), nloops);
    myproducer.join();

    myworker.Stop();
    std::cout << "all DONE" << std::endl;

    return 0;
}
